### PR TITLE
Add test case showing errors with Kotlin DSL

### PR DESCRIPTION
**Pull request by melix**
See #9249

### CODE REVIEW

**Comment by melix** [comment]
@bamboo @eskatos initially I wanted to fix #9249 with the assumption that it was just missing overloads. After writing the test above (actually migrating to a new test fixture), it shows that the situation is much worse than I expected. In particular, it seems that the problem is with the Kotlin compiler itself.

I have no idea how to fix this, any guidance would be appreciated.

**Comment by melix** [comment]
@bamboo any idea?

**Comment by eskatos** [comment]
@melix `@DslMarker` might help work the scopes as said in https://github.com/gradle/gradle/issues/9249#issuecomment-443018508

See https://kotlinlang.org/docs/reference/type-safe-builders.html#scope-control-dslmarker-since-11

Now that we have Kotlin in Gradle core it shouldn't be a problem to use it anymore.

**Comment by melix** [comment]
I tried to hack with `@DslMarker` but it doesn't work, or I didn't use it properly. There's almost no documentation unfortunately.

**Comment by melix** [comment]
There are 2 remaining tests failing with a suspicious error:

>   Line 35:     conf ("org:foo:1.0")
               ^ Overload resolution ambiguity: 
                   public final operator fun Configuration.invoke(dependencyNotation: Any): Dependency? defined in org.gradle.kotlin.dsl.DependencyHandlerScope
                   public final operator fun Configuration.invoke(dependencyNotation: Any): Dependency? defined in org.gradle.kotlin.dsl.DependencyHandlerScope
>
>  Line 37:         conf ("org:bar:1.1")
                   ^ Overload resolution ambiguity: 
                       public final operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint? defined in org.gradle.kotlin.dsl.DependencyConstraintHandlerScope
                       public final operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint? defined in org.gradle.kotlin.dsl.DependencyConstraintHandlerScope

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt [hunk 1]
```diff
@@ -43,6 +45,23 @@ private constructor(
     override val delegate: DependencyHandler
         get() = dependencies
 
+    @Deprecated(replaceWith = ReplaceWith("constraints"), message = "This method shouldn't be called because the most specific variant should be preferred by the Kotlin compiler", level = DeprecationLevel.HIDDEN)
```
Comment (bamboo) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt]: That's a creative usage of `HIDDEN`!

It does go at the heart of the problem. The wrong delegate type forced by the super method signature.

An alternative solution that might be longer term (with its own set of problems) would be to remove `DependencyHandler` from the `DependencyHandlerScope` type hierarchy altogether and keep a strict separation between API types and DSL types.

**Comment by bamboo** [comment]
> There are 2 remaining tests failing with a suspicious error:
> 
> > Line 35:     conf ("org:foo:1.0")
> > ^ Overload resolution ambiguity:
> > public final operator fun Configuration.invoke(dependencyNotation: Any): Dependency? defined in org.gradle.kotlin.dsl.DependencyHandlerScope
> > public final operator fun Configuration.invoke(dependencyNotation: Any): Dependency? defined in org.gradle.kotlin.dsl.DependencyHandlerScope
> > Line 37:         conf ("org:bar:1.1")
> > ^ Overload resolution ambiguity:
> > public final operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint? defined in org.gradle.kotlin.dsl.DependencyConstraintHandlerScope
> > public final operator fun Configuration.invoke(dependencyConstraintNotation: Any): DependencyConstraint? defined in org.gradle.kotlin.dsl.DependencyConstraintHandlerScope

It seems the `writeSpec` is emitting multiple `conf` variables in those cases.

**Comment by melix** [comment]
> It seems the writeSpec is emitting multiple conf variables in those cases.

That was indeed the problem, thanks!

**Comment by melix** [comment]
Candidate for backport to 6.2-rc-2 if there's one.

**Comment by eskatos** [comment]
With those changes the following works as expected:
```kotlin
dependencies {
  constraints {
    //...
  }
}
```

But the problem is still present when doing
```kotlin
dependencies.constraints {
  // ...
}
```

But it might be even more challenging to fix as this references `DependencyHandler` directly, there's no opportunity for the trick used on `DependencyHandlerScope`.

**Comment by melix** [comment]
Thanks @eskatos . Indeed this would have to be fixed too. I don't think it should prevent this PR from being merged though. The `dependencies.constraints` notation is still quite unusual.

**Comment by bamboo** [comment]
> But it might be even more challenging to fix as this references `DependencyHandler` directly, there's no opportunity for the trick used on `DependencyHandlerScope`.

Yes. I think the only safe long term solution here is to keep DSL types separate from the API types so their translation to Kotlin idioms can be completely controlled at the DSL level. In that world, `dependencies` wouldn't be a `DependencyHandler` but a `DependencyHandlerDsl` or something. You get the picture. This would also allow cleaning up the `Closure` overloads and other methods that we don't scripts to see.

**Comment by melix** [comment]
Yet it forces the creation of a parallel DSL that we shouldn't really need :/

**Comment by bamboo** [comment]
In any case, there's a lesson here, as long as we keep the current structure, we should avoid introducing these `modelElement(Action<ModelElementT>)` methods in the DSL types.

**Pull request merged**